File: quick_sort.py
import csv
import sys
import time

def read_csv(filename):
    with open(filename, newline='') as csvfile:
        reader = csv.reader(csvfile)
        return [(int(row[0]), row[1]) for row in reader]

def write_csv(filename, data):
    with open(filename, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        for item in data:
            writer.writerow([item[0], item[1]])

def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

def partition(arr, low, high):
    pivot = arr[high][0]  # still using last element
    i = low - 1
    for j in range(low, high):
        if arr[j][0] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def main():
    if len(sys.argv) != 2:
        print("Usage: python quick_sort.py dataset_n.csv")
        return

    input_file = sys.argv[1]
    data = read_csv(input_file)

    start_time = time.time()
    quick_sort(data, 0, len(data) - 1)
    end_time = time.time()

    output_file = f"quick_sort_{len(data)}.csv"
    write_csv(output_file, data)

    print(f"Sorted {len(data)} elements.")
    print(f"Output written to: {output_file}")
    print(f"Sorting time: {end_time - start_time:.6f} seconds")

if __name__ == "__main__":
    main()

File: quick_sort.java
import java.io.*;
import java.util.*;

public class quick_sort {

    static class Element {
        int number;
        String word;

        Element(int number, String word) {
            this.number = number;
            this.word = word;
        }
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java quick_sort dataset_n.csv");
            return;
        }

        String inputFile = args[0];
        List<Element> data;

        try {
            data = readCSV(inputFile);
        } catch (IOException e) {
            System.out.println("Error reading input file.");
            return;
        }

        long startTime = System.nanoTime();
        quickSort(data, 0, data.size() - 1);
        long endTime = System.nanoTime();

        String outputFile = "quick_sort_" + data.size() + ".csv";

        try {
            writeCSV(outputFile, data);
        } catch (IOException e) {
            System.out.println("Error writing output file.");
            return;
        }

        System.out.println("Sorted " + data.size() + " elements.");
        System.out.println("Output written to: " + outputFile);
        System.out.printf("Sorting time: %.6f seconds%n", (endTime - startTime) / 1e9);
    }

    static List<Element> readCSV(String filename) throws IOException {
        List<Element> list = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",", 2);
                list.add(new Element(Integer.parseInt(parts[0]), parts[1]));
            }
        }
        return list;
    }

    static void writeCSV(String filename, List<Element> data) throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (Element e : data) {
                writer.write(e.number + "," + e.word);
                writer.newLine();
            }
        }
    }

    static void quickSort(List<Element> arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    static int partition(List<Element> arr, int low, int high) {
        int pivot = arr.get(high).number;
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr.get(j).number < pivot) {
                i++;
                Collections.swap(arr, i, j);
            }
        }
        Collections.swap(arr, i + 1, high);
        return i + 1;
    }
}